Code/Framework/AzCore/Tests/Serialization/Json/MapSerializerTests.cpp

```diff
@@ -173,9 +173,29 @@ namespace JsonSerializationTests
         AZStd::shared_ptr<Map> CreatePartialDefaultInstance() override
         {
             auto instance = AZStd::shared_ptr<Map>(new Map{}, &Delete);
-            instance->emplace(AZStd::make_pair(aznew SimpleClass(), aznew SimpleClass(188, 188.0)));
-            instance->emplace(AZStd::make_pair(aznew SimpleClass(242, 242.0), aznew SimpleClass()));
-            instance->emplace(AZStd::make_pair(aznew SimpleClass(342, 342.0), aznew SimpleClass(388, 388.0)));
+            instance->emplace(AZStd::make_pair(aznew SimpleClass(), aznew SimpleClass()));
+            instance->emplace(AZStd::make_pair(aznew SimpleClass(), aznew SimpleClass()));
+            instance->emplace(AZStd::make_pair(aznew SimpleClass(), aznew SimpleClass()));
+            size_t i = 0;
+            for (auto& entry : *instance)
+            {
+                if (i == 0)
+                {
+                    *entry.first = {};
+                    *entry.second = {188, 188.0};
+                }
+                else if (i == 1)
+                {
+                    *entry.first = {242, 242.0};
+                    *entry.second = {};
+                }
+                else if (i == 2)
+                {
+                    *entry.first = {342, 342.0};
+                    *entry.second = {388, 388.0};
+                }
+                ++i;
+            }
```

Comment:
AZStd::map is an ordered container. When the key type for a map is a pointer, then the address that pointer points to is used for the sorting algorithm for where that item should appear in the map. When a map is serialized to json, it is merely iterated over in (sorted) order.

The json tests serialize a map out to a string, then compare it against a known string, so the order of the elements is important. That means that care must be taken when inserting elements into the map, that things are inserted in the correct order. For a map of pointers, that means that the memory address of the allocated objects is what determines the map element's order.

These tests were previously relying on the allocator to always return a higher address for later allocations, which is not a guarantee of the allocator. This fixes that error in the test by allocating the objects first, inserting them in the map so they are sorted by address, then traversing the map to set the values appropriately.

Reply:
Ok thanks for the clarification.

This means the test is invalid testing the order of elements, and it is something we should look into refactoring so that each JSON object stringified entry is compared instead of the entire stringified JSON object, but that is something that can be done in the future.